apps/prairielearn/src/pages/studentInstanceQuestionExam/studentInstanceQuestionExam.js

```diff
@@ -93,87 +95,140 @@ function processSubmission(req, res, callback) {
-  } else if (req.body.__action === 'attach_file') {
-    util.callbackify(studentInstanceQuestion.processFileUpload)(
-      req,
-      res,
-      function (err, variant_id) {
-        if (ERR(err, next)) return;
-        res.redirect(
-          res.locals.urlPrefix +
-            '/instance_question/' +
-            res.locals.instance_question.id +
-            '/?variant_id=' +
-            variant_id,
-        );
-      },
-    );
-  } else if (req.body.__action === 'attach_text') {
-    util.callbackify(studentInstanceQuestion.processTextUpload)(
-      req,
-      res,
-      function (err, variant_id) {
-        if (ERR(err, next)) return;
-        res.redirect(
-          res.locals.urlPrefix +
-            '/instance_question/' +
-            res.locals.instance_question.id +
-            '/?variant_id=' +
-            variant_id,
-        );
-      },
-    );
-  } else if (req.body.__action === 'delete_file') {
-    util.callbackify(studentInstanceQuestion.processDeleteFile)(
-      req,
-      res,
-      function (err, variant_id) {
+      }
+      if (req.body.__action === 'grade' && !res.locals.assessment.allow_real_time_grading) {
+        next(error.make(403, 'Real-time grading is not allowed for this assessment'));
+        return;
+      }
+      if (res.locals.assessment.group_work) {
+        const groupConfig = await groupAssessmentHelper.getGroupConfig(res.locals.assessment.id);
+        if (groupConfig.has_roles) {
+          const result = await groupAssessmentHelper.getQuestionPermissions(
+            res.locals.assessment_question.id,
+            res.locals.user.user_id,
+          );
+          // Users without the correct roles cannot submit
+          if (!result.can_submit) {
+            return next(
+              error.make(
+                403,
+                'Current group roles have no permission to submit this question',
+                res.locals,
+              ),
+            );
+          }
+
+          // Get group information
+          const groupId = await groupAssessmentHelper.getGroupId(
+            res.locals.assessment.id,
+            res.locals.user.user_id,
+          );
+          const groupInfo = await groupAssessmentHelper.getGroupInfo(groupId, groupConfig);
+          const { validationErrors, rolesAreBalanced, usersWithoutRoles } = groupInfo.rolesInfo;
+
+          // Prevent submissions if role configuration is invalid
+          if (validationErrors.length > 0 || !rolesAreBalanced || usersWithoutRoles.length > 0) {
+            return next(
+              error.make(
+                403,
+                'Current group role configuration is invalid. Submissions are disabled',
+                res.locals,
+              ),
+            );
+          }
+        }
+      }
+
+      await util.promisify(processSubmission)(req, res);
+      res.redirect(req.originalUrl);
+    } else if (req.body.__action === 'timeLimitFinish') {
+      // TODO: Does this need to be turned async?
```

Comment:
I don't believe we need to remove callback-based code in here, but I'm not entirely sure either.

Reply:
No, it shouldn't be necessary to do so; that can wait for a bigger refactor of the assessment module.